File: CosmoWhisper-Mac/CosmoWhisper/Managers/LicenseManager.swift
import Foundation
import AppKit

struct LicenseStatus: Codable {
    let tier: String
    let usageMinutes: Double
    let limitMinutes: Int
    let isOverLimit: Bool
}

class LicenseManager: ObservableObject {
    static let shared = LicenseManager()
    
    private let session = URLSession.shared
    
    func syncStatus() async -> Bool {
        let token = UserDefaults.standard.string(forKey: "licenseToken") ?? ""
        let backendUrl = UserDefaults.standard.string(forKey: "backendUrl") ?? "http://localhost:5000"
        
        guard !token.isEmpty else { return false }
        
        guard let url = URL(string: "\(backendUrl)/api/license/status") else { return false }
        
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        
        do {
            let (data, response) = try await session.data(for: request)
            if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 {
                let status = try JSONDecoder().decode(LicenseStatus.self, from: data)
                
                DispatchQueue.main.async {
                    UserDefaults.standard.set(status.tier, forKey: "userTier")
                    UserDefaults.standard.set(status.usageMinutes, forKey: "usageMinutes")
                    UserDefaults.standard.set(status.limitMinutes, forKey: "usageLimitMinutes")
                    UserDefaults.standard.set(!status.isOverLimit, forKey: "isAIUnlocked")
                    LogManager.shared.log("LICENSE: Sync successful. Tier: \(status.tier)")
                }
                return true
            }
        } catch {
            LogManager.shared.log("LICENSE: Sync failed: \(error.localizedDescription)")
        }
        
        return false
    }
    
    func reportUsage(durationSeconds: Double) async {
        let token = UserDefaults.standard.string(forKey: "licenseToken") ?? ""
        let backendUrl = UserDefaults.standard.string(forKey: "backendUrl") ?? "http://localhost:5000"
        
        guard !token.isEmpty else { return }
        
        guard let url = URL(string: "\(backendUrl)/api/license/report-usage") else { return }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body = ["durationMs": durationSeconds * 1000]
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        
        do {
            let (_, response) = try await session.data(for: request)
            if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 {
                let _ = await syncStatus()
            }
        } catch {
            LogManager.shared.log("LICENSE: Report failed: \(error.localizedDescription)")
        }
    }
}

File: CosmoWhisper-Windows/CosmoWhisper/Managers/LicenseManager.cs
using System;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text.Json;
using System.Threading.Tasks;
using System.Diagnostics;

namespace CosmoWhisper.Managers
{
    public class LicenseStatus
    {
        public string tier { get; set; } = "free";
        public double usageMinutes { get; set; }
        public int limitMinutes { get; set; }
        public bool isOverLimit { get; set; }
    }

    public class LicenseManager
    {
        public static LicenseManager Shared { get; } = new LicenseManager();
        private readonly HttpClient _httpClient;

        public LicenseManager()
        {
            _httpClient = new HttpClient();
            _httpClient.Timeout = TimeSpan.FromSeconds(10);
        }

        public async Task<bool> SyncStatusAsync()
        {
            var prefs = PreferenceManager.Shared.Preferences;
            if (string.IsNullOrEmpty(prefs.LicenseToken)) return false;

            try
            {
                _httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", prefs.LicenseToken);
                string url = $"{prefs.BackendUrl.TrimEnd('/')}/api/license/status";

                var response = await _httpClient.GetAsync(url);
                if (response.IsSuccessStatusCode)
                {
                    string json = await response.Content.ReadAsStringAsync();
                    var status = JsonSerializer.Deserialize<LicenseStatus>(json);

                    if (status != null)
                    {
                        prefs.UserTier = prefs.IsStoreVersion ? "free" : status.tier;
                        prefs.UsageMinutes = status.usageMinutes;
                        prefs.UsageLimitMinutes = prefs.IsStoreVersion ? 20 : status.limitMinutes;
                        prefs.IsAIUnlocked = !status.isOverLimit;

                        PreferenceManager.Shared.Save();
                        return true;
                    }
                }
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"[LICENSE] Sync failed: {ex.Message}");
            }

            return false;
        }

        public async Task ReportUsageAsync(double durationSeconds)
        {
            var prefs = PreferenceManager.Shared.Preferences;
            if (string.IsNullOrEmpty(prefs.LicenseToken)) return;

            try
            {
                _httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", prefs.LicenseToken);
                string url = $"{prefs.BackendUrl.TrimEnd('/')}/api/license/report-usage";

                var content = new StringContent(
                    JsonSerializer.Serialize(new { durationMs = durationSeconds * 1000 }),
                    System.Text.Encoding.UTF8,
                    "application/json"
                );

                var response = await _httpClient.PostAsync(url, content);
                if (response.IsSuccessStatusCode)
                {
                    // Refresh status after reporting
                    await SyncStatusAsync();
                }
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"[LICENSE] Report failed: {ex.Message}");
            }
        }
    }
}
